File: 872. Leaf-Similar Trees/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        return dfs(root1, new ArrayList<>()).equals(dfs(root2, new ArrayList<>()));
    }
    public List<Integer> dfs(TreeNode curr, List<Integer> sequence) {
        if (curr == null) return sequence;
        if (curr.left == null && curr.right == null) {
            sequence.add(curr.val);
        } else {
            dfs(curr.left, sequence);
            dfs(curr.right, sequence);
        }
        return sequence;
    }
}

File: 872. Leaf-Similar Trees/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        return self.dfs(root1, []) == self.dfs(root2, [])
    def dfs(self, curr, sequence):
        if curr == None: return sequence
        if curr.left == None and curr.right == None:
            sequence.append(curr.val)
        else:
            self.dfs(curr.left, sequence)
            self.dfs(curr.right, sequence)
        return sequence
